File: error-handling/error_handling.py
class AgeError(Exception):
    """Custom exception for invalid age values."""


class ConfigError(RuntimeError):
    """Custom unchecked-style exception (runtime error category)."""


def parse_age(text: str) -> int:
    try:
        age = int(text)
    except ValueError as exc:
        raise AgeError(f"Age must be a number: {text}") from exc

    if age < 0:
        raise AgeError("Age cannot be negative")

    return age


def load_config(data: dict) -> str:
    if "mode" not in data:
        raise ConfigError("Missing 'mode' in config")
    return data["mode"]


def checked_vs_unchecked_note() -> None:
    print("Python note: all exceptions are runtime; no checked exception enforcement.")


def main() -> None:
    print("Language: Python")

    # try/except/finally
    try:
        print("Parsed age:", parse_age("21"))
        print("Parsed age:", parse_age("oops"))
    except AgeError as err:
        print("Caught custom exception:", err)
    finally:
        print("Finally block: cleanup/logging runs always")

    # Another custom exception example
    try:
        mode = load_config({})
        print("Mode:", mode)
    except ConfigError as err:
        print("Caught runtime-style custom exception:", err)

    checked_vs_unchecked_note()


if __name__ == "__main__":
    main()

File: error-handling/errorHandling.js
class AgeError extends Error {
  constructor(message) {
    super(message);
    this.name = "AgeError";
  }
}

class ConfigError extends Error {
  constructor(message) {
    super(message);
    this.name = "ConfigError";
  }
}

function parseAge(text) {
  const age = Number(text);
  if (Number.isNaN(age)) {
    throw new AgeError(`Age must be a number: ${text}`);
  }
  if (age < 0) {
    throw new AgeError("Age cannot be negative");
  }
  return age;
}

function loadConfig(data) {
  if (!Object.prototype.hasOwnProperty.call(data, "mode")) {
    throw new ConfigError("Missing 'mode' in config");
  }
  return data.mode;
}

function checkedVsUncheckedNote() {
  console.log("JavaScript note: no checked exceptions; all throws are runtime.");
}

function main() {
  console.log("Language: JavaScript");

  try {
    console.log("Parsed age:", parseAge("21"));
    console.log("Parsed age:", parseAge("oops"));
  } catch (err) {
    if (err instanceof AgeError) {
      console.log("Caught custom exception:", err.message);
    } else {
      console.log("Caught unexpected error:", err);
    }
  } finally {
    console.log("Finally block: cleanup/logging runs always");
  }

  try {
    const mode = loadConfig({});
    console.log("Mode:", mode);
  } catch (err) {
    if (err instanceof ConfigError) {
      console.log("Caught runtime-style custom exception:", err.message);
    }
  }

  checkedVsUncheckedNote();
}

main();
